pkg/sysinfo/sysinfo.go

```diff
@@ -102,6 +117,128 @@ type cgroupPids struct {
 	PidsLimit bool
 }
 
+// ValidateHugetlb check whether hugetlb pagesize and limit legal
+func (c cgroupHugetlbInfo) ValidateHugetlb(pageSize string, limit uint64) (string, []string, error) {
+	var (
+		w   []string
+		err error
+	)
+	if pageSize != "" {
+		if err = isHugepageSizeValid(pageSize); err != nil {
+			return "", w, err
+		}
+	} else {
+		pageSize, err = c.GetDefaultHugepageSize()
+		if err != nil {
+			return "", w, fmt.Errorf("Failed to get system hugepage size")
+		}
+	}
+
+	warning, err := isHugeLimitValid(pageSize, limit)
+	w = append(w, warning...)
+	if err != nil {
+		return "", w, err
+	}
+
+	return pageSize, w, nil
+}
+
+// isHugeLimitValid check whether input hugetlb limit legal
+// it will check whether the limit size is times of size
+func isHugeLimitValid(size string, limit uint64) ([]string, error) {
+	var w []string
+	sizeInt, err := units.RAMInBytes(size)
+	if err != nil || sizeInt < 0 {
+		return w, fmt.Errorf("Invalid hugepage size:%s -- %s", size, err)
+	}
+	sizeUint := uint64(sizeInt)
+
+	if limit%sizeUint != 0 {
+		w = append(w, "Invalid hugetlb limit: should be multiple of huge page size")
+	}
+
+	return w, nil
+}
+
+// isHugepageSizeValid check whether input size legal
+// it will compare size with all system supported hugepage size
+func isHugepageSizeValid(size string) error {
+	hps, err := getHugepageSizes()
+	if err != nil {
+		return err
+	}
+
+	for _, hp := range hps {
+		if size == hp {
```

Comment:
Wonder if we can make this less brittle; do we know if case is consistent, and if `k` / `K` / `kB` / `KB` are not mixed? For example, `1g` may be a valid size, but the pagesize on the host may be named `/sys/fs/cgroup/hugetlb/hugetlb.1GB.limit_in_byte`.


If we _do_ want to validate, perhaps;
- Pass sizes as _bytes_ through the API, not strings
- Get all the defined sizes from the host (all of `hugetlb.XX.limit_in_bytes`)
- Create a map to map their actual size in bytes (e.g. `9223372036854771712` for 1GB) to _name_ of the size (`1GB`)

If the OCI spec expects the size to be specified as a string, we can map bytes to the expected string if needed

Reply:
I thought daemon host_config should take standard value as input, it's cli's job to parse user input into standard format.